mlflow/pyfunc/__init__.py

```diff
@@ -320,62 +433,192 @@ def predict(*args):
+                      environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults'],
+                            'dependencies': [
+                                'python=3.7.0',
+                                'cloudpickle==0.5.8'
+                            ]
+                        }
+
+    :param python_model: An instance of a subclass of :class:`~PythonModel`. This class will be
+                         serialized using the CloudPickle library. Any dependencies of the class
+                         should be included in one of the following locations:
+
+                            - The MLflow library.
+                            - Package(s) listed in the model's Conda environment, specified by
+                              the ``conda_env`` parameter.
+                            - One or more of the files specified by the ``code_path`` parameter.
+
+                        Note: If the class is imported from another module, as opposed to being
+                        defined in the ``__main__`` scope, the defining module should also be
+                        included in one of the listed locations.
+    :param artifacts: A dictionary containing ``<name, artifact_uri>`` entries. Remote artifact URIs
+                      will be resolved to absolute filesystem paths, producing a dictionary of
+                      ``<name, absolute_path>`` entries. ``python_model`` can reference these
+                      resolved entries as the ``artifacts`` property of the ``context`` attribute.
+                      For example, consider the following ``artifacts`` dictionary::
+
+                        {
+                            "my_file": "s3://my-bucket/path/to/my/file"
+                        }
+
+                      In this case, the ``"my_file"`` artifact will be downloaded from S3. The
+                      ``python_model`` can then refer to ``"my_file"`` as an absolute filesystem
+                      path via ``self.context.artifacts["my_file"]``.
+
+                      If *None*, no artifacts will be added to the model.
     """
-    if os.path.exists(dst_path):
-        raise Exception("Path '{}' already exists".format(dst_path))
-    os.makedirs(dst_path)
-    code = None
-    data = None
-    env = None
-
-    if data_path:
-        model_file = _copy_file_or_tree(src=data_path, dst=dst_path, dst_dir="data")
-        data = model_file
-
-    if code_path:
-        for path in code_path:
-            _copy_file_or_tree(src=path, dst=dst_path, dst_dir="code")
-        code = "code"
-
-    if conda_env:
-        shutil.copy(src=conda_env, dst=os.path.join(dst_path, "mlflow_env.yml"))
-        env = "mlflow_env.yml"
+    first_argument_set = {
+        "loader_module": loader_module,
+        "data_path": data_path,
+    }
+    second_argument_set = {
+        "artifacts": artifacts,
```

Comment:
Empirically, why could `artifacts` not be specified if I'm using a loader module? I understand the desire for either a tuple of (python file, entrypoint) vs python instance, since those map to how we construct the python object representation of the model, but artifacts seem orthogonal.

I wonder if during the MLflow 1.0 work we should consider revamping the naming of this API a bit -- `data_path` and `loader_module` do not seem related, but they both need to be specified together or not at all.